### ISSUE TITLE
ng command not recognized in eclipse preference

### ISSUE THREAD

**Issue by sayantansinha** [angelozerr/angular-eclipse#107]
After I add my angular-cli path on the Javascript->Typescript->Angular-CLI option under Eclipse Preference I get an error saying "The '/node_modules//bin/ng' is not a ng file". I am currently using macOS Sierra (v10.12.6) and using the same ng --version command provides the correct output listing down the version. I got this ng file after running npm install on my project's package.json which is referring to v1.0.2
Any help would be really appreciated.

**Comment by angelozerr** [comment]
> After I add my angular-cli path on the Javascript->Typescript->Angular-CLI option under Eclipse Preference 

I'm not sure, but I think this preference is not used with Angular Eclipse 1.3.0 since you can install ng cli in your wizard when you create a cli project. Hav you tried to create an new project with the wizard?

@Springrbua could you see if Javascript->Typescript->Angular-CLI option  is used? If not perhaps we should update the wizard?

**Comment by sayantansinha** [comment]
Let me give more background, I created this project using v1.2.0 and on a windows 7 os. Now, I am hust trying to use the same projecton a mac and I setup the ltes v1.3.0 of the plugin. I was not aware that this preference was not used anymore. So why is it there in the first place? If it is not used then how does the terminal recognizes the ng command? To further note, this is a company issued mac so I do not have admin privileges (going back to my previous issue where I highlighted the same problem for which the execute ng with file option was added by you), as a result of which I am not able to install ng globally as well. So, I am not trying to create a new project, I am trying to access an existing project and just trying to run it after setting it up.

**Comment by Springrbua** [comment]
@angelozerr The preferences are still in use. Actually the new wizard also just updates this preferences with the options you set.

@sayantansinha The field where you enter the ng location should already provide a default option, if you click on the little arrow on the left. This default option uses the $Project variable and links to the project specific ng installation.  
This value used to work for me, so you could try that.

**Comment by sayantansinha** [comment]
@Springrbua yes I tried that as well and it does not work on either of the platforms (windows or mac) for me. The preferences page shows an error on the top and does not allow me to got another preference popping a message saying this page has invalid options set.

**Comment by sayantansinha** [comment]
Finally it worked, I followed @Springrbua suggestion:

> EDIT:
> The global preferences are not used if you create a new project, as it always creates project specifc settings. If instead you import a project it will probably use the global settings.

I just turned on the project specific settings wherein the default worked

> ${project_loc:node_modules/.bin}

Also, had to make sure to delete any existing run configurations for the same project as changing the preferences and then running with the same run configuration did not pick up the latest preference (@Springrbua - shouldn't this be something that need to be fixed?)

Hope this might be helpful for someone in the future.

**Comment by Springrbua** [comment]
Thanks for the feedback!  
Probably the global preferences shouldn't validate ng installation, since it is possible to use project variables, which only exist in projects. @angelozerr what do you think about that?  

I have to take a look at the problem with the run configurations, I never had this bug but I'll try to reproduce it.  
Thanks for your patience and your detailed feedback :)

**Comment by angelozerr** [comment]
>  @angelozerr what do you think about that?

@Springrbua I'm very busy for the moment with Codelens, if you have time please try it and give me feedback, thanks!

**Comment by Springrbua** [comment]
@angelozerr I'll take a look at it as soon as possible

**Comment by angelozerr** [comment]
> Probably the global preferences shouldn't validate ng installation, since it is possible to use project variables

If I remember I had done that, to configure ng path for the whole project of the workspace (it's a kind of ng global installation). But if it's confusing, perhaps we should remove this preferences since now ng new project wizard gives the capability to use ng global installed angular-cli or a given angular-cli to install?

> @angelozerr I'll take a look at it as soon as possible

Thanks!

**Comment by Springrbua** [comment]
So you would remove the global settings? Or should I just add a "use global settings" option to the wizard and validate it only there?

**Comment by angelozerr** [comment]
> Or should I just add a "use global settings" option to the wizard and validate it only there?

Yes it was my idea, but if it's more complex to understand, perhaps we should remove this preferences, I don't know.

**Comment by Springrbua** [comment]
I guess we should keep the global preferences, because one could use a "global" cli-installation which is installed in some custom folder. This is currently not possible with the Wizard, since it lets you only choose global installation (installed with `npm install -g`) or install project specific cli.  
But I'll take a look at it to see whats better.

**Comment by angelozerr** [comment]
> I guess we should keep the global preferences, because one could use a "global" cli-installation which is installed in some custom folder. 

Exactly, it was my idea.

> This is currently not possible with the Wizard, since it lets you only choose global installation (installed with npm install -g) or install project specific cli.

Yes it's the problem. We should add it in the wizard or remove it from General / Preferences.

> But I'll take a look at it to see whats better.

Thanks!

**Comment by sayantansinha** [comment]
Adding my 2 cents, I think you guys also need to look into the fact that why the existing run configurations doesn't get updated with the latest changes I do to the preferences wizard. It always refers to the old one and you have to delete and reconfigure the run settings again to allow it to pick up the latest preferences.

**Comment by Springrbua** [comment]
@sayantansinha Yea of course, I'll also take a look at that. I'll try to reproduce the issue and I'll probably open a new Issue for it.  
Thanks again for your feedback!

**Comment by Springrbua** [comment]
@angelozerr My idea is to remove "use global CLI" and add "use global preferences" instead.  
So to use global cli, you simply need to select it in the global preferences. If you instead want to use a custom, non-global CLI, you can select that in the global preferences.  
The Project-Wizard then validates this CLI-installation.  
What do you think about that?

**Comment by angelozerr** [comment]
Yes it's a good idea.

**Comment by nitinkaveriappa** [comment]
I have the following error in Eclipse --> "The 'C:/Users/.../AppData/Roaming/npm/ng.cmd' is not a ng file."
I am not sure but I believe that it is related to the same issue.
I have attached screenshots of the same which might help you understand the issue better.
Is there a fix for this?
![capture](https://user-images.githubusercontent.com/6459566/36503041-d28eb5a6-1719-11e8-93b4-1d969d86829a.PNG)
![capture2](https://user-images.githubusercontent.com/6459566/36503042-d298c938-1719-11e8-9b75-56cef83bb72e.PNG)
![capture3](https://user-images.githubusercontent.com/6459566/36503043-d2a5c124-1719-11e8-84de-7bfec07275e8.PNG)

**Comment by angelozerr** [comment]
Perhaps it's again a bug with new Angular CLI version. Which version of Angular CLI are you using @nitinkaveriappa ?

**Comment by nitinkaveriappa** [comment]
@sayantansinha I am not able to create an Angular Project, nor am I able to change it in the Preference
@angelozerr I am using version 1.3.0

**Comment by angelozerr** [comment]
> @angelozerr I am using version 1.3.0

Are you sure it's the version Angular CLI. I think you are using last version of Angular CLI (1.5?)

I think you are using angular-eclipse 1.30, please install 1.4.0-SNAPSHOT with update site.

**Comment by nitinkaveriappa** [comment]
@angelozerr Sorry my mistake. Angular CLI version I am using is 1.7.0

**Comment by angelozerr** [comment]
> @angelozerr Sorry my mistake. Angular CLI version I am using is 1.7.0

Ok thanks, I think 1.7.0 have changed their "ng -version". I will see that when I will find time.

**Comment by l0v3w4v3** [comment]
@angelozerr is there anyway to fix @nitinkaveriappa problem yet?

**Comment by angelozerr** [comment]
I must just find time to study and fix the problem. But any contribution are welcome!

**Comment by trhsrw** [comment]
Dear @angelozerr, shouldn't you take the plugin from the Eclipse Marketplace as long as this issue is not fixed?

As far as I can see it, this issue has severity BROKEN und there is no workaround for Windows systems to fix it on the user side, or is there?

Selecting a custom installation of cli with a valid path to ng doesn't change anything to the error. Preferences dialog keeps saying ".../ng.cmd is not a ng file" regardless of which path is selected. Project specific settings can not be activated as long as this error persists ("The currently displayed page contains invalid values").

I am using the latest release from the marketplace, 1.3.0, CLI version is 6.0.8. Lower CLI versions force the same error.

**Comment by trhsrw** [comment]
@Springrbua I've tried that several times with different versions of CLI. Didn't work. It creates a project with just the node_modules folder in it. No src and e2e folders, no angular.json, package.json, tsconfig.json...

My point is not to blame you for your limited time. I think, many people could safe a lot of nerves and time if it was taken down from the marketplace as long as this is neither fixed nor a reasonable workaround is provided.

**Comment by angelozerr** [comment]
> Dear @angelozerr, shouldn't you take the plugin from the Eclipse Marketplace as long as this issue is not fixed?

is it working with snapshot update site? If yes, I think we should to create a new release.

**Comment by trhsrw** [comment]
What is the snapshot update site and where can I find it?

My Eclipse is currently broken by this error, because I can't close the Angular CLI preferences page anymore ("The currently displayed page contains invalid values").

**Comment by angelozerr** [comment]
> What is the snapshot update site and where can I find it?

https://github.com/angelozerr/angular-eclipse/wiki/Installation-Update-Site

**Comment by trhsrw** [comment]
The problem persists. I have checked the output of ng --version. That's kind of strange with ASCII art. I don't know, whether there is a version of the --version option without ascii art. You won't be able to fix this as long as they keep playing around with crap like that in their version output.

You might want to consider a boolean option (checkbox) in the preferences / the new project dialog to skip CLI validation. Or something like: "use ng.cmd anyway, I know, what I'm doing." This might help most people with a standard installation of angular cli.

**Comment by angelozerr** [comment]
Thanks for your feedback @trhsrw . I agree with you, this version search must perhaps be less restrictive. I must find time to study this problem and do something.

**Comment by JJN2** [comment]
PATH="$PATH:/usr/lib/node_modules//bin"  --> PATH="$PATH:/usr/lib/node_modules/@angular/cli/bin"
Sorry for that typo in example PATH setting

**Comment by JJN2** [comment]
ah.. special character which is parsed away from text.. well anyways.. set Angular path in there.

**Comment by Springrbua** [comment]
@JJN2 If you select `global cli`, you should have `angular-cli` installed globally, by using `npm install @angular/cli -g`. This should install it in the global `node_modules`, which should be part of the path by default.  
Basically, the plugin just executes `ng --version`, to check if a valid `angular-cli` version is available. If the `ng`-command is not part of the path, then it won't be able to execute it.  
Usually, it is a good idea to install `angukar-cli` globally, if you are developing applications with it, because otherwise you won't be able to execute all the `ng`-commands easily.

**Comment by TheCrazyT** [comment]
Problem is that the embedded nodejs is used inside the $PATH.
You probably need to setup an external path for nodejs-first because the newest "@angular/cli" is not compatible with the embedded nodejs that gets used.
The plugin uses embedded 6.9.4 by default wich won't work with the ng-command.

**Comment by garymazz** [comment]
I'm having the same problem 1 year later.

I've configured ng global 
Angular CLI: 8.3.6 is installed globally.
Directory C:\Program Files\nodejs is in the path..

I've also configured to use "installed" node.js 10.16.3 in directory C:\Program Files\nodejs  with same results. 

Is there a fix/work around for windows 10?

**Comment by trhsrw** [comment]
@garymazz: I ended up using the Eclipse WildWebDeveloper plugin (strongly recommended anyway) and the Angular CLI in Eclipse‘s terminal view. Works best for my purposes.
